### PR TITLE
Update margin between label and icon in Tab to better reflect Material specs

### PR DESCRIPTION
**Pull request by davidmartos96**
This PR improves the distance between the label and the icon in the Tab widget. 
I updated the margin to 2 pixels, taken from the Figma design page for Material 3. On Material 2 I left the default value of 10 pixels.

Related to #128696 (In particular, the distance between label and icon)

Here are some screenshots for comparison. I looked a bit into the other mentioned issue of the tab height not following the M3 spec. Flutter uses 72 and the spec uses 64. But because Tab is a PreferredSizeWidget, I don't think there is an easy way to provide a different size depending on `ThemeData.useMaterial3`, because there is no `BuildContext` available.
I provide a sample image for the 64 height as well for context on the linked issue, even though it's not part of the PR changes.

The screenshots are taken side by side with the image at: https://m3.material.io/components/tabs/guidelines

## Original

original

## New (tab height = 72, Flutter default for 8 years)

new_72

## New (tab height = 64, M3 spec)

new_64

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/tabs.dart [hunk 1]
```diff
@@ -159,11 +162,16 @@ class Tab extends StatelessWidget implements PreferredSizeWidget {
       label = icon!;
     } else {
       calculatedHeight = _kTextAndIconTabHeight;
+      final EdgeInsetsGeometry effectiveIconMargin = iconMargin ?? (
+        Theme.of(context).useMaterial3
+          ? const EdgeInsets.only(bottom: 2)
+          : const EdgeInsets.only(bottom: 10)
+      );
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:169]: These are typically defined by tokens that we import directly from the material design spec, you can see some of them below in the M2 and M3 defaults. 
I can see this value is listed in https://m3.material.io/components/tabs/specs, @TahaTesser does this mean there should be a token available for this now?

**Review by Piinks — COMMENTED** [review]
Hey @davidmartos96 thanks for contributing.

>  Flutter uses 72 and the spec uses 64. But because Tab is a PreferredSizeWidget, I don't think there is an easy way to provide a different size depending on ThemeData.useMaterial3, because there is no BuildContext available.

This is exactly right 💯 , and why is has not been updated before now. Since useMaterial3 is true by default now, we can likely update the preferred size to match the spec now. This would likely be best as a separate change though, it will likely take some time to land since there will be a lot of affected customers.

File: packages/flutter/lib/src/material/tabs.dart (hunk 1 above)
Comment (davidmartos96) [reply packages/flutter/lib/src/material/tabs.dart:169]: I knew about the tokens, but I wasn't able to find the margin in the spec. And I still can't... :sweat_smile:
Under which categories in the tree-view is it, and what is the ID of the token?

Comment (TahaTesser) [reply packages/flutter/lib/src/material/tabs.dart:169]: > I can see this value is listed in https://m3.material.io/components/tabs/specs, @TahaTesser does this mean there should be a token available for this now?

Yes, `TabBar` and many widgets are missing tokens in Flutter for a while such as https://github.com/flutter/flutter/issues/124965

It was discuss that we shouldn't let the missing tokens get in the way, we can hard-code values in the template class which can be replaced in the future when the tokens area available.

Comment (TahaTesser) [reply packages/flutter/lib/src/material/tabs.dart:169]: @davidmartos96 
Instead of hard-coding the value these values in the `effectiveIconMargin` variable when `iconMargin` is null. Please consider moving these values to the template class and place `iconMargin` as static value. This will keep the defaults values in a single place and make easier to update with tokens later.

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/material/tabs.dart [hunk 2]
```diff
@@ -2328,6 +2338,8 @@ class _TabsPrimaryDefaultsM3 extends TabBarTheme {
   TabAlignment? get tabAlignment => isScrollable ? TabAlignment.startOffset : TabAlignment.fill;
 
   static double indicatorWeight = 3.0;
+
+  static const EdgeInsetsGeometry iconMargin = EdgeInsets.only(bottom: 2);
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/tabs.dart:2342]: Does this also update `TabBar.secondary`?

File: packages/flutter/test/material/tabs_test.dart [hunk 3]
```diff
@@ -386,6 +386,37 @@ void main() {
...
+  testWidgets('Tab icon-label margin', (WidgetTester tester) async {
+    Widget buildApp({required bool useMaterial3}) {
+      final ThemeData theme = ThemeData(fontFamily: 'FlutterTest', useMaterial3: useMaterial3);
+      return MaterialApp(theme: theme, home: const Center(child: Material(
+        child: Tab(
+          icon: Icon(Icons.house),
+          text: 'x',
+        )
+      )));
+    }
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:398]: This is breaking multiple format rules.

File: packages/flutter/test/material/tabs_test.dart [hunk 4]
```diff
@@ -386,6 +386,37 @@ void main() {
...
+    // Material 2
+    await tester.pumpWidget(
+      buildApp(useMaterial3: false),
+    );
+    expect(getMarginBetweenIconAndLabel(), 10);
+
+    // Material 3
+    await tester.pumpWidget(
+      buildApp(useMaterial3: true),
+    );
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:415]: Material 2 and Material 3 require separate tests. Please search for other similar in the framework for reference.

File: packages/flutter/test/material/tabs_test.dart [hunk 5]
```diff
@@ -386,6 +386,37 @@ void main() {
...
+      final ThemeData theme = ThemeData(fontFamily: 'FlutterTest', useMaterial3: useMaterial3);
+      return MaterialApp(theme: theme, home: const Center(child: Material(
+        child: Tab(
+          icon: Icon(Icons.house),
+          text: 'x',
+        )
+      )));
+    }
+
+    double getMarginBetweenIconAndLabel() {
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:400]:
```suggestion
    double getIconMargin() {
```

File: packages/flutter/lib/src/material/tabs.dart [hunk 6]
```diff
@@ -159,11 +162,16 @@ class Tab extends StatelessWidget implements PreferredSizeWidget {
       label = icon!;
     } else {
       calculatedHeight = _kTextAndIconTabHeight;
+      final EdgeInsetsGeometry effectiveIconMargin = iconMargin ?? (
+        Theme.of(context).useMaterial3
+          ? _TabsPrimaryDefaultsM3.iconMargin
+          : _TabsDefaultsM2.iconMargin
+      );
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/tabs.dart:169]:
```suggestion
      final EdgeInsetsGeometry effectiveIconMargin = iconMargin ??
        (Theme.of(context).useMaterial3
          ? _TabsPrimaryDefaultsM3.iconMargin
          : _TabsDefaultsM2.iconMargin);
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/test/material/tabs_test.dart (hunk 4 above)
Comment (TahaTesser) [reply packages/flutter/test/material/tabs_test.dart:415]: Add a Separate Material 3 similar to one below, notice the title. (You can skip passing ThemeData in this test)

https://github.com/flutter/flutter/blob/5f17badf4118b474f0b97ee1afe0773ac92b1e4b/packages/flutter/test/material/filter_chip_test.dart#L239-L241

And a separate Material 2 test similar to one blow (For Material 2, set `useMaterial3` to `false` and pass a ThemeData in the `MaterialApp.theme` parameter directly)

https://github.com/flutter/flutter/blob/5f17badf4118b474f0b97ee1afe0773ac92b1e4b/packages/flutter/test/material/filter_chip_test.dart#L134-L140

File: packages/flutter/lib/src/material/tabs.dart (hunk 2 above)
Comment (davidmartos96) [reply packages/flutter/lib/src/material/tabs.dart:2342]: Yes it does, but the Tab (which is the one applying this margin), is not aware of primary/secondary. Only TabBar knows that.
I was doubting between adding it static to PrimaryDefaults or create a global const _tabIconMarginM3 with the template. But since you wrote Primary in the code snippet above I went with the first approach

File: packages/flutter/test/material/tabs_test.dart (hunk 3 above)
Comment (davidmartos96) [reply packages/flutter/test/material/tabs_test.dart]: I was trying to separate the important piece, which is the Tab. I went with a separate variable instead in the last commit

**Comment by davidmartos96** [comment]
@TahaTesser Kind of on-topic with m2 and m3. I'm not sure if you know the answer.
There are a few tests in `tabs_test.dart` written like this:

```dart
  testWidgets('Tab sizing - icon and text', (WidgetTester tester) async {
    final ThemeData theme = ThemeData(fontFamily: 'FlutterTest');
    final bool material3 = theme.useMaterial3;
    await tester.pumpWidget(
      MaterialApp(theme: theme, home: const Center(child: Material(child: Tab(icon: SizedBox(width: 10.0, height: 10.0), text: 'x')))),
    );
    expect(tester.renderObject<RenderParagraph>(find.byType(RichText)).text.style!.fontFamily, 'FlutterTest');
    expect(
      tester.getSize(find.byType(Tab)),
      material3 ? const Size(14.25, 72.0) : const Size(14.0, 72.0));
  });
```

Material 3 would always be true right now, so there are "hanging" expectations. Is this part of https://github.com/flutter/flutter/issues/139076 and just missing some refactoring?

**Comment by TahaTesser** [comment]
> @TahaTesser Kind of on-topic with m2 and m3. I'm not sure if you know the answer. There are a few tests in `tabs_test.dart` written like this:

Yep, some of the old tests isn't following format rules. I recommend following the correct format for new changes even if you duplicate old tests and only make changes related to the fix itself, not update unrelated tests.

File: packages/flutter/lib/src/material/tabs.dart [hunk 7]
```diff
@@ -132,7 +132,10 @@ class Tab extends StatelessWidget implements PreferredSizeWidget {
   ///
   /// Only useful when used in combination with [icon], and either one of
   /// [text] or [child] is non-null.
-  final EdgeInsetsGeometry iconMargin;
+  ///
+  /// If [ThemeData.useMaterial3] is true, the default value is 2 pixels of bottom margin.
+  /// Otherwise, the default value is 10 pixels of bottom margin.
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/tabs.dart:137]:
```suggestion
  /// Defaults to 2 pixels of bottom margin. If [ThemeData.useMaterial3] is false,
  /// then defaults to 10 pixels of bottom margin.
```

The `useMaterial3` flag is `true` default.

File: packages/flutter/test/material/tabs_test.dart [hunk 8]
```diff
@@ -386,6 +386,40 @@ void main() {
...
+
+    await tester.pumpWidget(buildApp());
+    expect(getIconMargin(), 10);
+  });
+
+  testWidgets('Material3 - Tab icon-label margin', (WidgetTester tester) async {
+    final ThemeData theme = ThemeData();
+    Widget buildApp() {
+      const Tab tab = Tab(icon: Icon(Icons.house), text: 'x');
+      return MaterialApp(theme: theme, home: const Center(child: Material(child: tab)));
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:410]:
```suggestion
      return MaterialApp(
        theme: theme,
        home: const Center(child: Material(child: tab),
      ));
```

Better readability

File: packages/flutter/test/material/tabs_test.dart [hunk 9]
```diff
@@ -386,6 +386,40 @@ void main() {
       material3 ? const Size(14.25, 72.0) : const Size(14.0, 72.0));
   });
 
+  testWidgets('Material2 - Tab icon-label margin', (WidgetTester tester) async {
+    final ThemeData theme = ThemeData(useMaterial3: false);
+    Widget buildApp() {
+      const Tab tab = Tab(icon: Icon(Icons.house), text: 'x');
+      return MaterialApp(theme: theme, home: const Center(child: Material(child: tab)));
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:393]:
```suggestion
      return MaterialApp(
        theme: theme,
        home: const Center(child: Material(child: tab),
      ));
```

File: packages/flutter/test/material/tabs_test.dart [hunk 10]
```diff
@@ -386,6 +386,40 @@ void main() {
       material3 ? const Size(14.25, 72.0) : const Size(14.0, 72.0));
   });
 
+  testWidgets('Material2 - Tab icon-label margin', (WidgetTester tester) async {
+    final ThemeData theme = ThemeData(useMaterial3: false);
+    Widget buildApp() {
+      const Tab tab = Tab(icon: Icon(Icons.house), text: 'x');
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:392]:
```suggestion

```
Does this have to be a variable? if it isn't reused, we can directly put this `Tab` in the `child` parameter and format the code accordingly

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/test/material/tabs_test.dart (hunk 10 above)
Comment (davidmartos96) [reply packages/flutter/test/material/tabs_test.dart:392]: Not necessarily, but I feel it reads better than the following, which is the test right above. The variable separates what's being tested.

```dart
await tester.pumpWidget(
    MaterialApp(theme: theme, home: const Center(child: Material(child: Tab(icon: SizedBox(width: 10.0, height: 10.0), child: Text('x'))))),
);
```

Would this be preferred?
```dart
return MaterialApp(
  theme: theme,
  home: const Center(child: Material(
    child: Tab(icon: Icon(Icons.house), text: 'x'),
  ),
));
```

Comment (TahaTesser) [reply packages/flutter/test/material/tabs_test.dart:392]: The second is better, add a line break for `text`.

Please read the [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) for this PR and future PRs.

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/test/material/tabs_test.dart (hunk 10 above)
Comment (davidmartos96) [reply packages/flutter/test/material/tabs_test.dart]: Done in a889db9

I did look into the style guide but I didn't find anything related to widget trees. Did I miss something? Apologies if that wasn't the expected format. I've seen in the tests that widget trees tend to be inlined to reduce the number of lines in the test files, unlike a "regular" Flutter app with dartfmt.

Thank you for the reviews on this and other PRs!

File: packages/flutter/lib/src/material/tabs.dart (hunk 2 above)
Comment (davidmartos96) [reply packages/flutter/lib/src/material/tabs.dart:2341]: How do you feel about the static field in this class which applies to both? Or do you have another suggestion?, considering that the Tab widget is not aware in which type of TabBar is positioned.

Comment (TahaTesser) [reply packages/flutter/lib/src/material/tabs.dart:2341]: I think `_TabsPrimaryDefaultsM3` static field is fine. This is just a temporary method (other similar classes do this too), eventually it will be replaced when we get update tokens in Flutter.

File: dev/tools/gen_defaults/lib/tabs_template.dart [hunk 11]
```diff
@@ -77,6 +77,8 @@ class _${blockName}PrimaryDefaultsM3 extends TabBarTheme {
   TabAlignment? get tabAlignment => isScrollable ? TabAlignment.startOffset : TabAlignment.fill;
 
   static double indicatorWeight = ${getToken('md.comp.primary-navigation-tab.active-indicator.height')};
+
+  static const EdgeInsetsGeometry iconMargin = EdgeInsets.only(bottom: 2);
```
Comment (TahaTesser) [inline dev/tools/gen_defaults/lib/tabs_template.dart:81]:
```suggestion

  // TODO(davidmartos96): This is taken from https://m3.material.io/components/tabs/specs
  // Update this when the token is available.
  static const EdgeInsetsGeometry iconMargin = EdgeInsets.only(bottom: 2);
```

Comment (TahaTesser) [reply dev/tools/gen_defaults/lib/tabs_template.dart:81]: Please run the script to update `tabs.dart` class  with this update

File: packages/flutter/test/material/tabs_test.dart [hunk 12]
```diff
@@ -386,6 +386,54 @@ void main() {
       material3 ? const Size(14.25, 72.0) : const Size(14.0, 72.0));
   });
 
+  testWidgets('Material2 - Tab icon-label margin', (WidgetTester tester) async {
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:389]:
```suggestion
  testWidgets('Material2 - Default Tab iconMargin', (WidgetTester tester) async {
```

File: packages/flutter/test/material/tabs_test.dart [hunk 13]
```diff
@@ -386,6 +386,54 @@ void main() {
...
+      final Rect iconRect = tester.getRect(find.byIcon(Icons.house));
+      final Rect labelRect = tester.getRect(find.text('x'));
+      return labelRect.top - iconRect.bottom;
+    }
+
+    await tester.pumpWidget(buildApp());
+    expect(getIconMargin(), 10);
+  });
+
+  testWidgets('Material3 - Tab icon-label margin', (WidgetTester tester) async {
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:413]:
```suggestion
  testWidgets('Material3 - Default Tab iconMargin', (WidgetTester tester) async {
```

File: packages/flutter/test/material/tabs_test.dart [hunk 14]
```diff
@@ -386,6 +386,54 @@ void main() {
...
+      return MaterialApp(
+        theme: theme,
+        home: const Center(child: Material(
+          child: Tab(
+            icon: Icon(Icons.house),
+            text: 'x',
+          ),
+        ),
+      ));
+    }
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:401]:
```suggestion
    await tester.pumpWidget(MaterialApp(
      theme: ThemeData(useMaterial3: false),
        home: const Material(
        child: Tab(
          icon: Icon(Icons.house),
          text: 'x',
        ),
      ),
    ));
```

We don't need variable if we're not testing any theme property.

File: packages/flutter/test/material/tabs_test.dart [hunk 15]
```diff
@@ -386,6 +386,54 @@ void main() {
...
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:425]:
```suggestion
    await tester.pumpWidget(const MaterialApp(
      home: Material(
        child: Tab(
          icon: Icon(Icons.house),
          text: 'x',
        ),
      ),
    ));
```

We don't need variable if we're not testing any theme property and since M3 is default, we don't even need to specify theme here.

File: packages/flutter/test/material/tabs_test.dart [hunk 16]
```diff
@@ -386,6 +386,54 @@ void main() {
...
+
+    double getIconMargin() {
+      final Rect iconRect = tester.getRect(find.byIcon(Icons.house));
+      final Rect labelRect = tester.getRect(find.text('x'));
+      return labelRect.top - iconRect.bottom;
+    }
+
+    await tester.pumpWidget(buildApp());
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:433]:
```suggestion
```
Pump the widget directly.

File: packages/flutter/test/material/tabs_test.dart [hunk 17]
```diff
@@ -386,6 +386,54 @@ void main() {
...
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:409]:
```suggestion
```
Pump the widget directly.

File: packages/flutter/test/material/tabs_test.dart [hunk 18]
```diff
@@ -386,6 +386,54 @@ void main() {
...
+    expect(getIconMargin(), 2);
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:434]:
```suggestion
    expect(getIconMargin(), equals(2));
```

Making sure expect value is exactly `2`, not a double.

File: packages/flutter/test/material/tabs_test.dart [hunk 19]
```diff
@@ -386,6 +386,54 @@ void main() {
...
+    }
+
+    double getIconMargin() {
+      final Rect iconRect = tester.getRect(find.byIcon(Icons.house));
+      final Rect labelRect = tester.getRect(find.text('x'));
+      return labelRect.top - iconRect.bottom;
+    }
+
+    await tester.pumpWidget(buildApp());
+    expect(getIconMargin(), 10);
```
Comment (TahaTesser) [inline packages/flutter/test/material/tabs_test.dart:410]:
```suggestion
    expect(getIconMargin(), equals(10));
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]
This is almost there! Thanks for working on this.

File: dev/tools/gen_defaults/lib/tabs_template.dart (hunk 11 above)
Comment (davidmartos96) [reply dev/tools/gen_defaults/lib/tabs_template.dart:81]: We might want to rephrase it to avoid confusion, as the token value doesn't exist in the spec. I took the value from the official Figma project for Material 3.

```
// TODO(davidmartos96): This value doesn't currently exist in https://m3.material.io/components/tabs/specs
// Update this when the token is available.
```

File: packages/flutter/test/material/tabs_test.dart (hunk 18 above)
Comment (davidmartos96) [reply packages/flutter/test/material/tabs_test.dart:434]: @TahaTesser  Does `equals` have extra checks? I've always thought the `equals` matcher was implicit when passing a value. 
I'm trying `expect(10.0, equals(10))` and the expectation succeeds.
Which `expect` case are you referring it could fail?
`expect(10.0, 10)` also succeeds.

File: dev/tools/gen_defaults/lib/tabs_template.dart (hunk 11 above)
Comment (davidmartos96) [reply dev/tools/gen_defaults/lib/tabs_template.dart]: Done in 2b809f6eec7214dfcc1f83701e31ec24a9566bf0

Comment (TahaTesser) [reply dev/tools/gen_defaults/lib/tabs_template.dart]: Break line after "in"

File: packages/flutter/test/material/tabs_test.dart (hunk 18 above)
Comment (TahaTesser) [reply packages/flutter/test/material/tabs_test.dart:425]: It succeed with direct value as well. `equals`  structure of the expected value.

In this case, we're checking a  `double` value against an  `int`, whatever reason this changes from `10.00001` we still need make sure it is `10`.  This is not necessary but good to have.

**Review by TahaTesser — APPROVED** [review]
LGTM!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you!
